### PR TITLE
Grpo loss

### PR DESCRIPTION
**Pull request by kashif**
## Summary
Adds the GRPO chunked loss

fixes issue #548 

## Testing Done

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: src/liger_kernel/chunked_loss/grpo_loss.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                ref_token_logprobs = ref_log_probs.gather(dim=-1, index=chosen_tokens.unsqueeze(-1)).squeeze(-1)
+        else:
+            ref_token_logprobs = chosen_token_logprobs.detach()
+
+        # Compute advantages per batch entry in a grouped fashion
+        mean_grouped_rewards = rewards.mean()  # [batch_size,]
+        std_grouped_rewards = rewards.std()  # [batch_size,]
+
+        # Calculate advantages using the same epsilon as in GRPOTrainer
+        advantages = (rewards - mean_grouped_rewards) / (std_grouped_rewards + 1e-8)
```
Comment (edbeeching) [inline src/liger_kernel/chunked_loss/grpo_loss.py:35]: I think we use 1e-4 in our implementation ?

Comment (kashif) [reply src/liger_kernel/chunked_loss/grpo_loss.py:35]: ah yes... i forgot to push that

File: test/chunked_loss/test_grpo_loss.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    input2 = _input.detach().clone().requires_grad_(True)
+
+    # Create attention mask with random padding [B*num_generations, T]
+    attention_mask = torch.ones(B * num_generations, T, device=device)
+    num_elements_to_mask = torch.randint(1, B * num_generations * T // 2, (1,)).item()
+    mask_indices = torch.randperm(B * num_generations * T)[:num_elements_to_mask]
+    attention_mask.view(-1)[mask_indices] = 0
+
+    # Create rewards with shape [B, num_generations]
+    rewards = torch.randn(B * num_generations, device=device, dtype=dtype)
```
Comment (edbeeching) [inline test/chunked_loss/test_grpo_loss.py:219]: If it is not too much work, would it be possible to test a scenario where the rewards are same same, i.e all 1s?

**Review by RichhLi — COMMENTED** [review]
Great work! Great use of comments

**Review by RichhLi — APPROVED** [review]

**Comment by kashif** [comment]
so there was a small change to the loss: https://github.com/huggingface/trl/pull/2881 which i will integrate here too

File: src/liger_kernel/chunked_loss/grpo_loss.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                ref_token_logprobs = ref_log_probs.gather(dim=-1, index=chosen_tokens.unsqueeze(-1)).squeeze(-1)
+        else:
+            ref_token_logprobs = chosen_token_logprobs.detach()
+
+        # Compute advantages per batch entry in a grouped fashion
+        mean_grouped_rewards = rewards.mean()  # [batch_size,]
+        std_grouped_rewards = rewards.std()  # [batch_size,]
+
+        # Calculate advantages using the same epsilon as in GRPOTrainer
+        advantages = (rewards - mean_grouped_rewards) / (std_grouped_rewards + 1e-4)
```
Comment (lancerts) [inline src/liger_kernel/chunked_loss/grpo_loss.py:35]: nit, we can define a eps var for 1e-4.

File: src/liger_kernel/chunked_loss/fused_linear_rlhf.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    @staticmethod
+    def forward(
+        ctx,
+        _input,
+        weight,
+        attention_mask,
+        rewards,
+        bias=None,
+        loss_fn=None,
+        num_generations=1,
```
Comment (lancerts) [inline src/liger_kernel/chunked_loss/fused_linear_rlhf.py:17]: nit, the default value
`num_generations: int = 4
`
is 4 below and we can make things consistent

**Review by lancerts — APPROVED** [review]

**Comment by lancerts** [comment]
Nice work!!
